src/components/PrimaryPanes/SourcesTree.js

```diff
@@ -115,28 +94,32 @@ class SourcesTree extends Component<Props, State> {
   }
 
   componentWillReceiveProps(nextProps) {
+    const {
+      projectRoot,
+      debuggeeUrl,
+      sources,
+      shownSource,
+      selectedSource
+    } = this.props;
+
+    const { uncollapsedTree, sourceTree } = this.state;
+
     if (
-      this.props.projectRoot !== nextProps.projectRoot ||
-      this.props.debuggeeUrl !== nextProps.debuggeeUrl
+      projectRoot != nextProps.projectRoot ||
+      debuggeeUrl != nextProps.debuggeeUrl ||
+      nextProps.sources.size === 0
     ) {
-      // Recreate tree because the sort order changed
-      this.setState(
-        createTree(
-          nextProps.sources,
-          nextProps.debuggeeUrl,
-          nextProps.projectRoot
-        )
+      // early recreate tree because of changes
```

Comment:
feels a little vague, can this be abit more detailed?

Reply:
Sure.